### PR TITLE
[SPARK-48056][CONNECT][FOLLOW-UP] Scala Client re-execute plan if a SESSION_NOT_FOUND error is raised and no partial response was received

### PR DESCRIPTION
**Pull request by ghost**
### What changes were proposed in this pull request?

This change lets a Scala Spark Connect client reattempt execution of a plan when it receives a SESSION_NOT_FOUND error from the Spark Connect service if it has not received any partial responses.

This is a Scala version of the previous fix of the same issue - https://github.com/apache/spark/pull/46297.

### Why are the changes needed?

Spark Connect clients often get a spurious error from the Spark Connect service if the service is busy or the network is congested. This error leads to a situation where the client immediately attempts to reattach without the service being aware of the client; this leads to a query failure.

### Does this PR introduce _any_ user-facing change?

Prevoiusly, a Scala Spark Connect client would fail with the error code "INVALID_HANDLE.SESSION_NOT_FOUND" in the very first attempt to make a request to the service, but with this change, the client will automatically retry.

### How was this patch tested?

Attached unit test.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by juliuszsompolski — APPROVED** [review]

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/SparkConnectClientSuite.scala [hunk 1]
```diff
@@ -530,6 +530,24 @@ class SparkConnectClientSuite extends ConnectFunSuite with BeforeAndAfterEach {
...
+    startDummyServer(0)
+    client = SparkConnectClient
+      .builder()
+      .connectionString(s"sc://localhost:${server.getPort}")
+      .enableReattachableExecute()
+      .build()
+    service.errorToThrowOnExecute = Some(new StatusRuntimeException(
+      Status.INTERNAL.withDescription("INVALID_HANDLE.SESSION_NOT_FOUND")))
+
+    val plan = buildPlan("select * from range(10000000)")
```
Comment (juliuszsompolski) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/SparkConnectClientSuite.scala:543]: nit: this test probably doesn't need such a large query, doing just a range(100) may maybe shave 50-100ms or sth, which is not a lot but adds up in thousands of tests...

Comment (changgyoopark-db) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/SparkConnectClientSuite.scala:543]: totally makes sense, I'll adjust the test a little bit.

Comment (juliuszsompolski) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/SparkConnectClientSuite.scala:543]: edit: even better, make it just a `select 1`, no need to launch any Spark jobs

**Comment by changgyoopark-db** [comment]
@HyukjinKwon Hi! Could you review and possible merge this PR? It is a follow-up change of https://github.com/apache/spark/pull/46297 which was merged in May.

**Review by zhengruifeng — APPROVED** [review]

**Pull request closed without merging**